src/bvh/iter.rs

```diff
@@ -0,0 +1,377 @@
+use crate::aabb::Bounded;
+use crate::bvh::{BVHNode, BVH};
+use crate::ray::Ray;
+
+/// Iterator traverse a BVH without memory allocations
+pub struct BVHIterator<'a, Shape: Bounded> {
+    bvh: &'a BVH,
+    ray: &'a Ray,
+    shapes: &'a [Shape],
+    stack: [usize; 32],
+    node_index: usize,
+    stack_size: usize,
+    has_node: bool,
+}
+
+impl<'a, Shape: Bounded> BVHIterator<'a, Shape> {
+    /// Creates a new BVHIterator
+    pub fn new(bvh: &'a BVH, ray: &'a Ray, shapes: &'a [Shape]) -> Self {
+        BVHIterator {
+            /// Reference to the BVH to traverse
+            bvh: bvh,
+
+            /// Reference to the input ray
+            ray: ray,
+
+            /// Reference to the input shapes array
+            shapes: shapes,
+
+            /// Traversal stack. 4 billion items seems enough?
+            stack: [0; 32],
+
+            /// Position of the iterator in bvh.nodes
+            node_index: 0,
+
+            /// Size of the traversal stack
+            stack_size: 0,
+
+            /// Whether or not we have a valid node (or leaf)
+            has_node: true,
+        }
+    }
+
+    /// Test if stack is empty.
+    fn is_stack_empty(&self) -> bool {
+        return self.stack_size == 0;
+    }
+
+    /// Push node onto stack. Not guarded against overflow.
+    fn stack_push(&mut self, node: usize) {
+        self.stack[self.stack_size] = node;
+        self.stack_size += 1;
+    }
+
+    /// Pop the stack and return the node. Not guarded against underflow.
+    fn stack_pop(&mut self) -> usize {
+        self.stack_size -= 1;
+        return self.stack[self.stack_size];
+    }
+
+    /// Attempt to move to the left child of the current node.
```

Comment:
"Attempt" implies that it can fail but it seems it can't really fail or at least not that I can see. Please document what "attempt" means here and what an unsuccessful action looks like in this function and `move_right`.